MdeModulePkg/Core/Dxe/Mem/Page.c

```diff
@@ -1741,6 +1741,36 @@ CoreFreePages (
 {
   EFI_STATUS       Status;
   EFI_MEMORY_TYPE  MemoryType;
+  UINT64           Attributes;
+
+  // check if this memory is returned to the core as RW at a minimum. If the memory attribute protocol is not installed,
+  // then we assume that the memory is RW by default and continue to free it.
+  if (gMemoryAttributeProtocol != NULL) {
+    Status = gMemoryAttributeProtocol->GetMemoryAttributes (
+                                         gMemoryAttributeProtocol,
+                                         Memory,
+                                         EFI_PAGES_TO_SIZE (NumberOfPages),
+                                         &Attributes
+                                         );
+
+    // if we failed to get the attributes, or if the memory is read-only or read-protected,
+    // then we leak the memory and return success. This is to avoid a known bootloader bug where it free memory that is

```

Comment:
How is this a bootloader bug? Does the memory attribute protocol definition include any guidance as to whether a user of the protocol should restore a region's permission attributes before freeing it? This seems like the responsibility of the DXE core to me.

Reply:
Yes, please update the comment. In general, referring to a 'known bootloader bug' without providing any details is highly likely to cause confusion in the future, and so it would be much better to add an annotation here that we failed to specify the expected behavior when we defined the protocol, and as a result, we cannot make any assumption whatsoever about the permission attributes of pages being handed back to the page allocator.